### PR TITLE
Close #1533 create api promotion coupon code

### LINKED ISSUE #1533: Create api promotion coupon code

### PR DESCRIPTION
**Pull request by Sreyleak-Deth**
### Add the code line below to store_decorator.rb because it undefined method 'default_notification_image' 
```
 base.has_one :default_notification_image, class_name: 'SpreeCmCommissioner::FeatureImage', dependent: :destroy, as: :viewable
 base.accepts_nested_attributes_for :default_notification_image, reject_if: :all_blank
 
```

### BaseUrl:  
```
{{BASE_URL}}/api/v2/storefront/user_promotion?include=default_store

```
```json
{
    "data": [
        {
            "id": "121",
            "type": "promotion",
            "attributes": {
                "name": "20% off for New Users",
                "description": "spen 4$,  valid for a minimun order of $5, Expiration: Promotion subject to change or termination.",
                "code": "ADFFD+",
                "expires_at": "2024-06-28T12:00:00.000+07:00",
                "starts_at": "2024-06-27T12:00:00.000+07:00",
                "usage_limit": 2,
                "promotion_category_id": 2,
                "created_at": "2024-06-27T15:17:40.740+07:00",
                "updated_at": "2024-07-12T18:02:50.847+07:00",
                "public_metadata": {}
            },
            "relationships": {
                "default_store": {
                    "data": {
                        "id": "1",
                        "type": "store"
                    }
                }
            }
        },
        {
            "id": "120",
            "type": "promotion",
            "attributes": {
                "name": "20% Off for New Users",
                "description": "Min spend $5",
                "code": "NEW2BM+",
                "expires_at": "2024-06-28T12:00:00.000+07:00",
                "starts_at": "2024-06-27T12:00:00.000+07:00",
                "usage_limit": 10,
                "promotion_category_id": 4,
                "created_at": "2024-06-27T11:25:05.990+07:00",
                "updated_at": "2024-07-02T10:37:40.651+07:00",
                "public_metadata": {}
            },
            "relationships": {
                "default_store": {
                    "data": {
                        "id": "1",
                        "type": "store"
                    }
                }
            }
        },
        {
            "id": "118",
            "type": "promotion",
            "attributes": {
                "name": "PO: 5$ off by Run with Me",
                "description": "Min Spen 2.5$",
                "code": "POMC+",
                "expires_at": "2024-06-30T12:00:00.000+07:00",
                "starts_at": "2024-06-28T12:00:00.000+07:00",
                "usage_limit": 10,
                "promotion_category_id": 2,
                "created_at": "2024-06-13T20:28:12.727+07:00",
                "updated_at": "2024-07-02T10:38:40.855+07:00",
                "public_metadata": {}
            },
            "relationships": {
                "default_store": {
                    "data": {
                        "id": "1",
                        "type": "store"
                    }
                }
            }
        },
        {
            "id": "109",
            "type": "promotion",
            "attributes": {
                "name": "Student Ticket Promotion 10% off",
                "description": "MIn Spen 0.5$",
                "code": "STUDENT",
                "expires_at": "2024-07-19T12:00:00.000+07:00",
                "starts_at": "2024-07-03T12:00:00.000+07:00",
                "usage_limit": null,
                "promotion_category_id": null,
                "created_at": "2024-02-02T12:35:48.177+07:00",
                "updated_at": "2024-07-03T09:53:08.884+07:00",
                "public_metadata": {}
            },
            "relationships": {
                "default_store": {
                    "data": {
                        "id": "1",
                        "type": "store"
                    }
                }
            }
        }
    ],
    "included": [
        {
            "id": "1",
            "type": "store",
            "attributes": {
                "term_and_condition_promotion": "<p>- 20% Off for New Users: Get a 20% discount on your first purchase.</p>\r\n<p>- Use Code: <strong>NEW2BM+</strong>: Enter code \"<strong>NEW2BM+</strong>\" at checkout.</p>\r\n<p>- Minimum Order <strong>$10</strong>: Minimum purchase of <strong>$10</strong> required.</p>\r\n<p>- Eligibility: Offer valid for new users only.</p>\r\n<p>- One-Time Use: Code can only be used once per user.</p>\r\n<p>- No Stackable Offers: Cannot be combined with other promotions.</p>\r\n<p>- Expiration: Promotion subject to change or termination.</p>\r\n<p>- Non-Transferable: Code has no cash value, non-transferable.</p>"
            }
        }
    ],
    "meta": {
        "count": 4,
        "total_count": 4,
        "total_pages": 1
    },
    "links": {
        "self": "http://127.0.0.1:4000//api/v2/storefront/user_promotion?include=default_store",
        "next": "http://127.0.0.1:4000/api/v2/storefront/user_promotion?include=default_store&page=1",
        "prev": "http://127.0.0.1:4000/api/v2/storefront/user_promotion?include=default_store&page=1",
        "last": "http://127.0.0.1:4000/api/v2/storefront/user_promotion?include=default_store&page=1",
        "first": "http://127.0.0.1:4000/api/v2/storefront/user_promotion?include=default_store&page=1"
    }
}
```

### CODE REVIEW

**Review by channainfo — CHANGES_REQUESTED** [review]

File: app/controllers/spree/api/v2/storefront/promo_codes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            Spree::Promotion.where(id: promotion_ids)
+          end
+
+          private
+
+          def fetch_user_specific_promotion_ids(user)
+            Spree::PromotionRuleUser
+              .joins(promotion_rule: :promotion)
+              .where(user_id: user.id)
+              .pluck('spree_promotions.id')
```
Comment (channainfo) [inline app/controllers/spree/api/v2/storefront/promo_codes_controller.rb:20]: This method can not be paginated and cause trouble if the dataset is huge

**Review by channainfo — CHANGES_REQUESTED** [review]

File: app/controllers/spree/api/v2/storefront/promo_codes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+          def collection
+            promotion_ids = fetch_user_specific_promotion_ids(spree_current_user)
+
+            Spree::Promotion.where(id: promotion_ids)
+          end
+
+          private
+
+          def fetch_user_specific_promotion_ids(user)
+            Spree::PromotionRuleUser
```
[the remark this answers is not in the document]
Comment (Sreyleak-Deth) [reply app/controllers/spree/api/v2/storefront/promo_codes_controller.rb:17]: > Create a custom relation instead. we prefer accessing promotion from spree_current_user.xxx instead.

Noted Bong. I will resolve it again.

File: app/controllers/spree/api/v2/storefront/promo_codes_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
+module Spree
+  module Api
+    module V2
+      module Storefront
+        class PromoCodesController < ::Spree::Api::V2::ResourceController
+          before_action :require_spree_current_user
+
+          def collection
+            @collection ||= spree_current_user.promotions
```
Comment (channainfo) [inline app/controllers/spree/api/v2/storefront/promo_codes_controller.rb:9]: sort it by newest first.

**Review by channainfo — CHANGES_REQUESTED** [review]

File: app/controllers/spree/api/v2/storefront/promo_codes_controller.rb
Comment (Thornsopanha) [inline app/controllers/spree/api/v2/storefront/promo_codes_controller.rb:1]: created_at change it to id

File: app/controllers/spree/api/v2/storefront/promo_codes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+          before_action :require_spree_current_user
+
+          def collection
+            @collection ||= spree_current_user.promotions.order(created_at: :desc)
+          end
+
+          private
+
+          def collection_serializer
+            Spree::Api::V2::Platform::PromotionSerializer
```
Comment (channainfo) [inline app/controllers/spree/api/v2/storefront/promo_codes_controller.rb:15]: Do not use this serializer

**Review by channainfo — CHANGES_REQUESTED** [review]

File: app/serializers/spree/v2/storefront/user_promotion_serializer.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+module Spree
+  module V2
+    module Storefront
+      class UserPromotionSerializer < BaseSerializer
+        attributes :name, :description, :code, :expires_at, :starts_at, :usage_limit, :type, :promotion_category_id, :created_at, :updated_at,
+                   :public_metadata
+
+        attribute :default_store do |_object|
```
Comment (channainfo) [inline app/serializers/spree/v2/storefront/user_promotion_serializer.rb:8]: Don't use this, use the association and include it instead.

**Review by channainfo — CHANGES_REQUESTED** [review]

File: app/serializers/spree/v2/storefront/user_promotion_serializer.rb (hunk 5 above)
Comment (Sreyleak-Deth) [reply app/serializers/spree/v2/storefront/user_promotion_serializer.rb:8]: Noted bong, I will update it

File: app/models/spree_cm_commissioner/promotion_decorator.rb [hunk 6]
```diff
@@ -2,10 +2,17 @@ module SpreeCmCommissioner
...
       base.attr_accessor :auto_apply
-
       base.before_validation :setup_auto_apply_promotion, if: :auto_apply?
     end
 
+    def default_store_id
+      stores.default&.id
+    end
+
+    def default_store
```
Comment (channainfo) [inline app/models/spree_cm_commissioner/promotion_decorator.rb:12]: Create a custom relation Instead.

File: app/serializers/spree/v2/storefront/user_promotion_serializer.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+module Spree
+  module V2
+    module Storefront
+      class UserPromotionSerializer < BaseSerializer
+        set_type :promotion
+
+        attributes :name, :description, :code, :expires_at, :starts_at, :usage_limit, :promotion_category_id, :created_at, :updated_at,
+                   :public_metadata
+
+        has_one :default_store, serializer: Spree::V2::Storefront::StoreSerializer
```
Comment (channainfo) [inline app/serializers/spree/v2/storefront/user_promotion_serializer.rb:10]: Create a custom relation instead otherwise it will cause n+1 query

**Review by channainfo — CHANGES_REQUESTED** [review]

**Pull request merged**